### PR TITLE
Refactoring for precomputations

### PR DESCRIPTION
**Pull request by aeyno**
This PR refactors the dynamic function's part of Puffin to use `dyn EvaluatedTerm` instead of `dyn Any` as input and output of the dynamic functions :
```rust
pub trait DynamicFunction<PT: ProtocolTypes>: Fn(&Vec<Box<dyn EvaluatedTerm<PT>>>) -> Result<Box<dyn EvaluatedTerm<PT>>, FnError> { }
```

This will allow calling `extract_knowledge()` on any concretized term, thus allowing to implement precomputations in a future PR.
The other effect of this is that it allows the compiler to check that all types that might be manipulated during the concretization of a term are implementing `EvaluatedTerm`.

This PR also introduce the `ProtocolTypes` trait which contains the protocol's `Matcher` type as an associated type and a function returning the signature of the protocol. This type is used to parametrize many of the types in Puffin to give them access to the matcher and signature of the protocol.

Most of the added impl of EvaluatedTerm for TLS and SSH in the PR are dummy knowledge extractions (mainly extracting self). In the future, we might want to be a bit more thorough and check that this new extractable knowledge can in fact be used during fuzzing.

### CODE REVIEW

File: puffin/src/trace.rs [hunk 1]
```diff
@@ -201,15 +197,15 @@ impl<PB: ProtocolBehavior> KnowledgeStore<PB> {
...
+        query: &Query<PT::Matcher>,
+    ) -> Option<&(dyn VariableData<PT>)> {
         let query_type_id: TypeId = query_type_shape.into();
 
-        let mut possibilities: Vec<Knowledge<PB::Matcher>> = self
+        let mut possibilities: Vec<Knowledge<PT>> = self
             .raw_knowledge
             .iter()
-            .filter(|raw| (query.source.is_none() || query.source.as_ref().unwrap() == &raw.source))
+            .filter(|raw| (query.source == None || query.source.as_ref().unwrap() == &raw.source))
```
Comment (michaelmera) [inline puffin/src/trace.rs:208]: Keep `.is_none()`. It's [best practice](https://rust-lang.github.io/rust-clippy/master/index.html#/partialeq_to_none).

**Review by michaelmera — CHANGES_REQUESTED** [review]
I marked some small mistakes that need to be fixed before merging this PR, but it's mainly the resultant of the big merge/rebase you had to do. I'm okay with the conceptual changes that this PR implements.

Also, the first commit of the PR should be removed: it's reverted by a following commit and is not part of this PR.

File: puffin/src/algebra/atoms.rs [hunk 2]
```diff
@@ -121,26 +125,26 @@ impl Function {
...
     ) {
         self.fn_container.shape = shape;
         self.fn_container.dynamic_fn = dynamic_fn;
     }
 }
 
-impl fmt::Display for Function {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+impl<PT: ProtocolTypes> fmt::Display for Function<PT> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
```
[the remark this answers is not in the document]
Comment (aeyno) [reply puffin/src/algebra/atoms.rs]: Done, but why do we need to use a path to use Formatter instead of just importing it ?

**Review by LCBH — APPROVED** [review]
LGTM

**Review by michaelmera — APPROVED** [review]

**Pull request merged**